### ISSUE TITLE
False positive `invalid-argument-type` with `Union` of `Callable[Concatenate[...]]`

### ISSUE THREAD

**Issue by SebbyLaw** [astral-sh/ty#3151]
## Summary

playground link
https://play.ty.dev/a4dea51f-75f7-49c1-b600-c001cb532ce2

ty 0.0.26 reports a false positive `invalid-argument-type` when a function is decorated with a decorator whose parameter type is a `Union` of two `Callable[Concatenate[...]]` forms with different arities. ty generates an intersection type instead of matching the function against each union arm independently.

In the wild, this affects every use of discord.py's `hybrid_command()`, `hybrid_group()`, and `HybridGroup.command()` decorators.

## Reproduction with discord.py

```
uv init repro && cd repro
uv add discord-py
uv add --dev ty
```

`repro.py`:
```python
from discord.ext.commands import Bot, Cog, Context, hybrid_command, hybrid_group


class MyCog(Cog):
    @hybrid_command()
    async def my_command(self, ctx: Context[Bot]) -> None:
        pass

    @hybrid_group()
    async def my_group(self, ctx: Context[Bot]) -> None:
        pass

    @my_group.command("sub")
    async def my_subcommand(self, ctx: Context[Bot]) -> None:
        pass
```

```
$ uv run ty check repro.py        # 0.0.26: error[invalid-argument-type]
$ uv run --with ty==0.0.25 ty check repro.py  # 0.0.25: All checks passed!
```

## Minimal reproduction (no dependencies)

`repro_no_deps.py`:
```python
from collections.abc import Callable, Coroutine
from typing import Any, Concatenate, Generic, ParamSpec, TypeVar, Union

P = ParamSpec("P")
T = TypeVar("T")
A = TypeVar("A")


class Ctx(Generic[T]):
    pass


B = TypeVar("B", bound=Ctx[Any])

Callback = Union[
    Callable[Concatenate[A, B, P], Coroutine[Any, Any, T]],
    Callable[Concatenate[B, P], Coroutine[Any, Any, T]],
]


def decorator() -> Callable[[Callback[A, B, P, T]], None]:
    def inner(func: Callback[A, B, P, T]) -> None:
        pass

    return inner


@decorator()
async def foo(x: int, y: Ctx[int]) -> None:
    pass
```

```
$ ty check repro_no_deps.py  # error[invalid-argument-type]
```

## Error output (ty 0.0.26)

```
error[invalid-argument-type]: Argument is incorrect
  --> repro_no_deps.py:28:1
   |
28 | @decorator()
   | ^^^^^^^^^^^^ Expected `((int, Ctx[int] | (int & Top[Ctx[Any]]), /) -> Coroutine[Any, Any, None]) | ((Ctx[int] | (int & Top[Ctx[Any]]), /) -> Coroutine[Any, Any, None])`, found `def foo(x: int, y: Ctx[int]) -> CoroutineType[Any, Any, None]`
29 | async def foo(x: int, y: Ctx[int]) -> None:
30 |     pass
```

## Analysis

`Callback` is a union of two callable forms with different `Concatenate` arities:

```python
Callback = Union[
    Callable[Concatenate[A, B, P], ...],   # arm 1: 2+ params before P
    Callable[Concatenate[B, P], ...],      # arm 2: 1 param before P
]
```

When matching `def foo(x: int, y: Ctx[int]) -> None`:

- **Arm 1** matches: `A = int`, `B = Ctx[int]`, `P = ()`, `T = None`
- **Arm 2** does not match: would need `B = int`, but `B` is bounded to `Ctx[Any]`

ty should accept the function since it matches arm 1. Instead, it unifies constraints across both arms simultaneously, producing the intersection type `int & Top[Ctx[Any]]`.

## Versions

- ty: 0.0.26 (940305127 2026-03-26)
- discord.py: 2.4.0+ (tested with 2.7.1)
- Python: 3.12+

### Version

ty 0.0.26 (940305127 2026-03-26)

**Comment by SebbyLaw** [comment]
after further investigation it looks like this only reproduces when the class used in the TypeVar bound is defined with legacy Generic syntax (`class Ctx(Generic[T])`)

**Comment by carljm** [comment]
Thanks for the report! I think the root cause of this is #2799. Here's a simpler repro that eliminates some distracting factors.

```py
from collections.abc import Callable
from typing import Concatenate, Generic, ParamSpec, TypeVar, Union

P = ParamSpec("P")
T = TypeVar("T")
A = TypeVar("A")
B = TypeVar("B")

Callback = Union[
    Callable[Concatenate[A, B, P], T],
    Callable[Concatenate[B, P], T],
]

def decorator() -> Callable[[Callback[A, B, P, T]], None]:
    def inner(func: Callback[A, B, P, T]) -> None:
        pass

    return inner

@decorator()
def foo(x: int, y: str) -> None:
    pass
```

When applying the decorator, we solve the type of `B` as `str | int`, because we try to match both `x` and `y` against `B`, and we naively union the types, which is wrong since we are matching in contravariant position. And `foo` is not assignable to either branch of the `Callback` union if `B` is solved to `str | int`, because callback arguments are in contravariant position.

The typevar bound with a generic type introduces some extra wrinkles (we intersect with the top materialization of the upper bound, so we get the type `(Ctx[int] | int) & Top[Ctx[Any]]`, which distributes over the union and simplifies to `Ctx[int] | (int & Top[Ctx[Any]]`. But the root cause is still the incorrect unioning.

> after further investigation it looks like this only reproduces when the class used in the TypeVar bound is defined with legacy Generic syntax (`class Ctx(Generic[T])`)

This is only because variance is inferred for PEP 695 generics, and with an empty class body (no use of the typevar) we infer the variance as "bivariant", making all specializations of `Ctx` equivalent to each other. If you introduce any use of `T` in the body of `Ctx`, making the typevar infer as covariant or invariant, the issue reproduces again.

**Comment by SebbyLaw** [comment]
thanks for the explanation! and sorry about the duplicate issue. didn't realize I did that

**Comment by carljm** [comment]
no worries! thanks for reporting!
